### PR TITLE
Task run implementation

### PR DESCRIPTION
**Pull request by schaable**
Part of https://github.com/NomicFoundation/hardhat/issues/5229

### CODE REVIEW

File: v-next/core/src/internal/tasks/resolved-task.ts [hunk 1]
```diff
@@ -69,24 +77,190 @@ export class ResolvedTask implements Task {
...
+    ): param is PositionalTaskParameter => "isVariadic" in param;
+    const isVariadic = isPositionalParameter(parameter) && parameter.isVariadic;
+
+    // check if the value is valid for the parameter type
+    if (!isParameterValueValid(parameter.parameterType, value, isVariadic)) {
+      throw new HardhatError(
+        HardhatError.ERRORS.ARGUMENTS.INVALID_VALUE_FOR_TYPE,
+        {
+          value,
+          name: parameter.name,
```
Comment (alcuadrado) [inline v-next/core/src/internal/tasks/resolved-task.ts:204]: Again, i think adding the task name would be valuable

File: v-next/core/src/internal/tasks/resolved-task.ts [hunk 2]
```diff
@@ -69,24 +77,190 @@ export class ResolvedTask implements Task {
...
+   * @throws HardhatError if the parameter is required and doesn't have a value.
+   */
+  #validateRequiredParameter(
+    parameter: TaskParameter,
+    value: ParameterValue | ParameterValue[],
+  ) {
+    if (parameter.defaultValue === undefined && value === undefined) {
+      throw new HardhatError(
+        HardhatError.ERRORS.ARGUMENTS.MISSING_VALUE_FOR_PARAMETER,
+        {
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply v-next/core/src/internal/tasks/resolved-task.ts:169]: The reason is that tasks can call other tasks, making this harder to debug otherwise

File: v-next/core/src/internal/tasks/resolved-task.ts [hunk 3]
```diff
@@ -69,24 +77,190 @@ export class ResolvedTask implements Task {
...
+   *
+   * @throws HardhatError if extra arguments were provided. The error message
+   * includes the name of the first extra argument.
+   */
+  #validateExtraArguments(providedArgumentNames: Set<string>) {
+    if (providedArgumentNames.size > 0) {
+      throw new HardhatError(
+        HardhatError.ERRORS.ARGUMENTS.UNRECOGNIZED_NAMED_PARAM,
+        {
+          parameter: [...providedArgumentNames][0],
```
Comment (alcuadrado) [inline v-next/core/src/internal/tasks/resolved-task.ts:222]: Same

File: v-next/core/src/internal/tasks/resolved-task.ts [hunk 4]
```diff
@@ -69,24 +77,190 @@ export class ResolvedTask implements Task {
...
+    } catch (error) {
+      ensureError(error);
+      throw new HardhatError(
+        HardhatError.ERRORS.TASK_DEFINITIONS.INVALID_ACTION_URL,
+        {
+          action: actionFileUrl,
+          task: formatTaskId(taskId),
+        },
+        error,
+      );
```
Comment (alcuadrado) [inline v-next/core/src/internal/tasks/resolved-task.ts:251]: To be addressed in a separate PR, but here is where you need to "validate" the plugin's installation.

File: v-next/core/src/internal/tasks/resolved-task.ts [hunk 5]
```diff
@@ -69,24 +77,190 @@ export class ResolvedTask implements Task {
...
+    // At this point, the set should be empty as all the task parameters have
+    // been processed. If there are any extra parameters, an error is thrown
+    this.#validateExtraArguments(providedArgumentNames);
+
+    const next = async (
+      nextTaskArguments: TaskArguments,
+      currentIndex = this.actions.length - 1,
+    ): Promise<any> => {
+      // The first action may be empty if the task was originally an empty task
+      const currentAction = this.actions[currentIndex].action ?? (() => {});
```
Comment (alcuadrado) [inline v-next/core/src/internal/tasks/resolved-task.ts:128]: Isn't this handled by `isEmpty` at the top? It think so.

If that's the case, I think it's better to have an `assertHardhatInvariant` and let it crash if the invariant doesn't hold.

File: v-next/core/src/internal/tasks/resolved-task.ts [hunk 6]
```diff
@@ -69,24 +77,190 @@ export class ResolvedTask implements Task {
...
+    // skip type validation for optional parameters with undefined value
+    if (value === undefined && parameter.defaultValue !== undefined) {
+      return;
+    }
+
+    // check if the parameter is variadic
+    const isPositionalParameter = (
+      param: TaskParameter,
+    ): param is PositionalTaskParameter => "isVariadic" in param;
+    const isVariadic = isPositionalParameter(parameter) && parameter.isVariadic;
```
[the remark this answers is not in the document]
Comment (schaable) [reply v-next/core/src/internal/tasks/resolved-task.ts:196]: if you do that, `isVariadic` resolves to `unknown` :(

File: v-next/core/src/internal/tasks/resolved-task.ts (hunk 4 above)
Comment (schaable) [reply v-next/core/src/internal/tasks/resolved-task.ts:251]: Yes, I've added it to the main task issue description.

File: v-next/core/src/internal/tasks/resolved-task.ts (hunk 5 above)
Comment (schaable) [reply v-next/core/src/internal/tasks/resolved-task.ts:128]: This is a different scenario. The `isEmpty` check at the top addresses empty tasks with just one action (the one from the empty task). However, you can override the empty task and add an action to it. In this case, you need to define an empty function to avoid errors when the override action calls `runSuper`. Does that make sense? There's a test that may clarify the scenario.

File: v-next/core/src/internal/tasks/resolved-task.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (schaable) [reply v-next/core/src/internal/tasks/resolved-task.ts:169]: I'll have to duplicate the descriptor because this one it's used in `v-next/hardhat/src/internal/cli/main.ts`, where there's no `taskId` to provide.

File: v-next/core/src/internal/tasks/resolved-task.ts (hunk 1 above)
Comment (schaable) [reply v-next/core/src/internal/tasks/resolved-task.ts:204]: Same here

File: v-next/core/src/internal/tasks/resolved-task.ts (hunk 3 above)
Comment (schaable) [reply v-next/core/src/internal/tasks/resolved-task.ts:222]: And same here 😅

File: v-next/core/src/internal/tasks/resolved-task.ts (hunk 5 above)
Comment (alcuadrado) [reply v-next/core/src/internal/tasks/resolved-task.ts:128]: Makes sense

File: v-next/core/test/internal/tasks/task-manager.ts [hunk 7]
```diff
@@ -875,4 +877,661 @@ describe("TaskManagerImplementation", () => {
...
+      assert.equal(overrideTaskRun, true);
+    });
+
+    it("should run a task with an action url", async () => {
+      const actionUrl = pathToFileURL(
+        path.resolve(
+          import.meta.dirname,
+          "./fixture-projects/file-actions/action-fn.js",
+        ),
+      ).toString();
```
Comment (alcuadrado) [inline v-next/core/test/internal/tasks/task-manager.ts:1173]: FYI:

`import.meta.resolve("./fixture-projects/file-actions/action-fn.js")` is equivalent

**Review by alcuadrado — APPROVED** [review]
I left a few comments, but LGTM

File: v-next/core/test/internal/tasks/task-manager.ts (hunk 7 above)
Comment (schaable) [reply v-next/core/test/internal/tasks/task-manager.ts:1173]: Nice, I'll change it just to have it as a reference for the future.
